plugins/basic/nexus-repository-maven/src/main/java/org/sonatype/nexus/repository/maven/internal/maven2/metadata/MetadataRebuilder.java

```diff
@@ -0,0 +1,418 @@
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.Maps;
+import com.orientechnologies.orient.core.command.OCommandResultListener;
+import com.orientechnologies.orient.core.db.ODatabaseRecordThreadLocal;
+import com.orientechnologies.orient.core.db.document.ODatabaseDocumentTx;
+import com.orientechnologies.orient.core.id.ORID;
+import com.orientechnologies.orient.core.metadata.schema.OType;
+import com.orientechnologies.orient.core.record.impl.ODocument;
+import com.orientechnologies.orient.core.sql.query.OSQLAsynchQuery;
+import org.w3c.dom.Document;
+import org.w3c.dom.NodeList;
+import org.xml.sax.SAXException;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkNotNull;
+import static com.google.common.base.Preconditions.checkState;
+
+/**
+ * Maven 2 repository metadata re-builder.
+ *
+ * @since 3.0
+ */
+@Singleton
+@Named
+@ThreadSafe
+public class MetadataRebuilder
+    extends ComponentSupport
+{
+  private final BucketEntityAdapter bucketEntityAdapter;
+
+  @Inject
+  public MetadataRebuilder(final BucketEntityAdapter bucketEntityAdapter)
+  {
+    this.bucketEntityAdapter = checkNotNull(bucketEntityAdapter);
+  }
+
+  /**
+   * Rebuilds/updates Maven metadata.
+   *
+   * @param repository  The repository who's metadata needs rebuild (Maven2 format, Hosted type only).
+   * @param update      if {@code true}, updates existing metadata, otherwise overwrites them with newly generated
+   *                    ones.
+   * @param groupId     scope the work to given groupId.
+   * @param artifactId  scope the work to given artifactId (groupId must be given).
+   * @param baseVersion scope the work to given baseVersion (groupId and artifactId must ge given).
+   */
+  public void rebuild(final Repository repository,
+                      final boolean update,
+                      @Nullable final String groupId,
+                      @Nullable final String artifactId,
+                      @Nullable final String baseVersion)
+  {
+    checkNotNull(repository);
+    final StringBuilder sql = new StringBuilder(
+        "SELECT " +
+            "group as groupId, " +
+            "name as artifactId, " +
+            "set(attributes.maven2.baseVersion) as baseVersions " +
+            "FROM component WHERE bucket=:bucket"
+    );
+    final Map<String, Object> sqlParams = Maps.newHashMap();
+    if (!Strings.isNullOrEmpty(groupId)) {
+      sql.append(" and group=:groupId");
+      sqlParams.put("groupId", groupId);
+      if (!Strings.isNullOrEmpty(artifactId)) {
+        sql.append(" and name=:artifactId");
+        sqlParams.put("artifactId", artifactId);
+        if (!Strings.isNullOrEmpty(baseVersion)) {
+          sql.append(" and attributes.maven2.baseVersion=:baseVersion");
+          sqlParams.put("baseVersion", baseVersion);
+        }
+      }
+    }
+    sql.append(" GROUP BY group, name LIMIT=-1");
```

Comment:
I created the SQL in console, and console has default limit 20, and at the end I just C+P the SQL into here :) Removing it.